File: functional/queens.rb
#!/usr/bin/env ruby

def nqueens(size)
    queen = lambda do |n|
        if n <= 0 then
            [[]]
        else
           queen.call(n-1).map{|x| (1..size).select{|y| !x.include?(y) && !x.map.with_index{|e,i| (y-e).abs-(x.length-i)}.include?(0)}.map{|y| x+[y]}}.flatten(1)
        end
    end
    queen.call(size)
end

puts nqueens(ARGV[0].to_i).map{|x| x.join(' ')}

File: functional/queens.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Queens
{
	public static void Main(string[] args) {
		Func<int, IEnumerable<IEnumerable<int>>> nqueens = size => {
			Func<int, IEnumerable<IEnumerable<int>>> queen = null;
			queen = n =>
				n <= 0 ?
				new [] {new int [] {}} :
				queen(n-1).SelectMany(a => Enumerable.Range(1, size).Where(b => !a.Contains(b) && !a.Select((c, i) => a.Count()-i-Math.Abs(b-c)).Contains(0)), (a, b) => a.Concat(new [] {b}));
			return queen(size);
		};
		Console.WriteLine(string.Join("\n", nqueens(int.Parse(args[0])).Select(x => string.Join(" ", x))));
	}
}
